server/mqtt.go

```diff
@@ -3451,6 +3506,40 @@ func (c *client) mqttProcessPubAck(pi uint16) {
 	sess.mu.Unlock()
 }
 
+// Process a PUBREL packet (QoS2, acting as Receiver).
+// Updates the session's pending list and sends an ACK to JS.
+//
+// Runs from the client's readLoop.
+// No lock held on entry.
+func (c *client) mqttProcessPubRel(pi uint16, trace bool) {
+	sess := c.mqtt.sess
+	if sess == nil {
+		return
+	}
+
+	isPending := false
+	sess.mu.Lock()
+	if sess.c != c {
+		sess.mu.Unlock()
+		return
+	}
+
+	if len(sess.pendingQOS2Received) > 0 {
```

Comment:
Again, since this is not persisted, if the server were to crash after sending the PUBREC and the sender sends the PUBREL, we would not reply with a PUBCOMP here (as we should). So I wonder if the send of the PUBCOMP should not be done regardless if the PI is found to be pending or not (if that is the case, then it could be simplified to a delete (not lookup and delete).

Reply:
ditto